File: module_test_1/src/main/org/grynko/nazar/java/BusStop.java
package main.org.grynko.nazar.java;

import lombok.SneakyThrows;

import java.util.Random;
import java.util.concurrent.Semaphore;

public class BusStop {

    private final Semaphore semaphore;
    private final String name;

    public BusStop(int permits, String name) {
        this.semaphore = new Semaphore(permits, true);
        this.name = name;
    }

    @SneakyThrows
    public void stop() {
        this.semaphore.acquire();

        String busName = Thread.currentThread().getName();

        System.out.printf("%s has stopped at the bus stop '%s'.\n", busName, name);
        randomSleep();
        System.out.printf("%s has left the bus stop '%s'.\n", busName, name);

        this.semaphore.release();
    }

    @SneakyThrows
    private void randomSleep() {
        Random random = new Random();
        Thread.sleep(random.nextInt(200) + 300);
    }

}

File: module_test_1/src/main/org/grynko/nazar/go/bus_stop/bus_stop.go
package bus_stop

import (
	"Lab_3/module_test/semaphore"
	"fmt"
	"math/rand"
	"time"
)

type BusStop struct {
	sem  semaphore.Semaphore
	name string
}

func Init(permit int, name string) *BusStop {
	bs := new(BusStop)

	bs.sem = semaphore.New(permit)
	bs.name = name

	return bs
}

func (busStop *BusStop) Stop(busName string) {
	busStop.sem.Acquire()

	println(fmt.Sprintf("%s has stopped at the bus stop '%s'.", busStop.name, busName))
	time.Sleep(time.Duration(rand.Intn(200)+300) * time.Millisecond)
	println(fmt.Sprintf("%s has left the bus stop '%s'", busStop.name, busName))

	busStop.sem.Release()
}
